gateway/metrics/exporter.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Alex Ellis 2017. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package metrics
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"log"
+
+	"github.com/openfaas/faas/gateway/requests"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Exporter is a prometheus exporter
+type Exporter struct {
+	metricOptions MetricOptions
+	services      []requests.Function
+}
+
+// NewExporter creates a new exporter for the OpenFaaS gateway metrics
+func NewExporter(options MetricOptions) *Exporter {
+	return &Exporter{
+		metricOptions: options,
+		services:      []requests.Function{},
+	}
+}
+
+// Describe is to describe the metrics for Prometheus
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	e.metricOptions.GatewayFunctionInvocation.Describe(ch)
+	e.metricOptions.GatewayFunctionsHistogram.Describe(ch)
+	e.metricOptions.ServiceReplicasCounter.Describe(ch)
+}
+
+// Collect collects data to be consumed by prometheus
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	e.metricOptions.GatewayFunctionInvocation.Collect(ch)
+	e.metricOptions.GatewayFunctionsHistogram.Collect(ch)
+
+	e.metricOptions.ServiceReplicasCounter.Reset()
```

Comment:
I think the service replica metric should be a [gauge](https://prometheus.io/docs/concepts/metric_types/#gauge) not a counter.

Reply:
Thank you Stefan. Actually, it is a Gauge.

https://github.com/openfaas/faas/blob/d701d03cdcd0dff4999e636ff8a75fc23927eb04/gateway/metrics/metrics.go#L16

Haven't changed the name but maybe we should rename it for clarity?

/cc @alexellis

Reply:
Renaming it would be great